### PR TITLE
Course file manager

### PR DESCRIPTION
**Pull request by nikke234**
This PR is against @superseacat's current PR (which can be merged first, then I'll just change the merge target to `test`), with a small refactor that moves course file related stuff out of `APlusProject` to a separate `CourseFileManager` class. I haven't updated tests or removed obsolete code, as I'm not sure if this is worth pursuing further, what do you think?

# Testing
<table>
  <tr>
    <th>unit</th>
    <th>integration</th>
    <th>manual</th>
  </tr>
  <tr>
    <td>
      <ul>
        <li>- [x] new <b>unit</b> tests created</li>
        <li>- [ ] all <b>unit</b> tests pass</li>
      </ul>
    </td>
    <td>
      <ul>
        <li>- [ ] new <b>integration</b> tests created</li>
        <li>- [x] all <b>integration</b> tests pass</li>
      </ul>
    </td>
    <td>
      <ul>
        <li>- [ ] <b>manual</b> testing went well</li>
      </ul>
    </td>
  </tr>
</table>  
  
# Have you updated the [TESTING.md](/TESTING.md) or other relevant documentation?

- [ ] Yes
- [ ] Not yet. I will do it next.
- [x] Not relevant

### CODE REVIEW

File: src/test/java/fi/aalto/cs/apluscourses/presentation/MainViewModelUpdaterTest.java [hunk 1]
```diff
@@ -102,7 +102,7 @@ public void testGetUpdatableModulesWithNullCourseReturnEmpty() {
   }
 
   @Test
-  public void testGetUpdatableModulesWithGetCourseFileModuleIdsThrowingReturnsEmpty()
+  public void disabledTestGetUpdatableModulesWithGetCourseFileModuleIdsThrowingReturnsEmpty()
```
Comment (superseacat) [inline src/test/java/fi/aalto/cs/apluscourses/presentation/MainViewModelUpdaterTest.java:105]: did you mean [`@Ignore`](https://github.com/junit-team/junit4/wiki/Ignoring-tests) or ...?

File: src/main/java/fi/aalto/cs/apluscourses/intellij/utils/CourseFileManager.java [hunk 2]
```diff
@@ -0,0 +1,159 @@
...
+   * course file for already exists (even with a different URL), this method loads the existing
+   * course file.
+   * @param project      The project for which the course file is created.
+   * @param courseUrl    The URL that gets added to the course file (if it doesn't exist yet).
+   * @throws IOException If an IO error occurs.
+   */
+  public synchronized void createAndLoad(@NotNull Project project, @NotNull URL courseUrl)
+      throws IOException {
+    courseFile = Paths
+        .get(project.getBasePath(), Project.DIRECTORY_STORE_FOLDER, COURSE_FILE_NAME)
```
Comment (superseacat) [inline src/main/java/fi/aalto/cs/apluscourses/intellij/utils/CourseFileManager.java:50]: might be `null` (IDEA is complaining)

File: src/main/java/fi/aalto/cs/apluscourses/intellij/actions/CourseProjectAction.java [hunk 3]
```diff
@@ -165,7 +165,7 @@ private Course tryInitializeCourse(@NotNull Project project, @NotNull URL course
   private boolean tryCreateCourseFile(@NotNull Project project, @NotNull URL courseUrl) {
```
Comment (superseacat) [inline src/main/java/fi/aalto/cs/apluscourses/intellij/actions/CourseProjectAction.java:165]: **Sonar** slightly dislikes it with `SonarLint: Refactor this method to not always return the same value. [+2 locations] ` and it's probably for a reason :wink: 

I strongly encourage everyone to install a [SonarLint IDEA plugin](https://plugins.jetbrains.com/plugin/7973-sonarlint) and connect it with our instance, so to get such notification even before pushing (there is a check 'before commit' which is superuseful)

File: src/test/java/fi/aalto/cs/apluscourses/utils/CourseFileManagerTest.java [hunk 4]
```diff
@@ -0,0 +1,135 @@
...
+import java.util.concurrent.atomic.AtomicBoolean;
+import org.apache.commons.io.FileUtils;
+import org.json.JSONObject;
+import org.junit.Assert;
+import org.junit.Test;
+
+public class CourseFileManagerTest {
+
+  @Test
+  public void testAddEntryForModule() throws IOException, InterruptedException {
```
Comment (superseacat) [inline src/test/java/fi/aalto/cs/apluscourses/utils/CourseFileManagerTest.java:31]: nit: could the internals be grouped and possibly separated into smaller helper methods for better readability?

File: src/test/java/fi/aalto/cs/apluscourses/utils/CourseFileManagerTest.java [hunk 5]
```diff
@@ -0,0 +1,135 @@
+package fi.aalto.cs.apluscourses.intellij.utils;
```
Comment (superseacat) [inline src/test/java/fi/aalto/cs/apluscourses/utils/CourseFileManagerTest.java:1]: some weird issue with the package name (IJ's notification)

File: src/main/java/fi/aalto/cs/apluscourses/intellij/utils/CourseFileManager.java [hunk 6]
```diff
@@ -0,0 +1,159 @@
...
+   * @param project The project from which the course file is loaded.
+   * @return {@code true} if the course file was successfully loaded, {@code false} if the course
+   *         file doesn't exist.
+   * @throws IOException   If an IO error occurs while reading the course file.
+   * @throws JSONException If the course file contains malformed JSON.
+   */
+  public synchronized boolean load(@NotNull Project project) throws IOException {
+    courseFile = Paths
+        .get(project.getBasePath(), Project.DIRECTORY_STORE_FOLDER, COURSE_FILE_NAME)
+        .toFile();
```
Comment (superseacat) [inline src/main/java/fi/aalto/cs/apluscourses/intellij/utils/CourseFileManager.java:74]: here is a minor code overlap with `createAndLoad(...)` (lines 49-51)

btw, would it make sense, that if the course file does not exist, then `createAndLoad(...)` will delegate to `load(...)`?

File: src/main/java/fi/aalto/cs/apluscourses/intellij/utils/CourseFileManager.java [hunk 7]
```diff
@@ -0,0 +1,159 @@
...
+  /**
+   * Attempts to load the course file corresponding to the given project. Returns {@code false} if
+   * the course file doesn't exist, {@code true} otherwise.
+   * @param project The project from which the course file is loaded.
+   * @return {@code true} if the course file was successfully loaded, {@code false} if the course
+   *         file doesn't exist.
+   * @throws IOException   If an IO error occurs while reading the course file.
+   * @throws JSONException If the course file contains malformed JSON.
+   */
+  public synchronized boolean load(@NotNull Project project) throws IOException {
```
Comment (superseacat) [inline src/main/java/fi/aalto/cs/apluscourses/intellij/utils/CourseFileManager.java:71]: I'd be glad to meet `@Test` for this method :smiley:

File: src/main/java/fi/aalto/cs/apluscourses/intellij/utils/CourseFileManager.java [hunk 8]
```diff
@@ -0,0 +1,159 @@
...
+
+  /**
+   * Attempts to create a course file and load it for the given project with the given URL. If a
+   * course file for already exists (even with a different URL), this method loads the existing
+   * course file.
+   * @param project      The project for which the course file is created.
+   * @param courseUrl    The URL that gets added to the course file (if it doesn't exist yet).
+   * @throws IOException If an IO error occurs.
+   */
+  public synchronized void createAndLoad(@NotNull Project project, @NotNull URL courseUrl)
```
Comment (superseacat) [inline src/main/java/fi/aalto/cs/apluscourses/intellij/utils/CourseFileManager.java:47]: > I'd be glad to meet `@Test` for this method

and this :smiling_face_with_three_hearts:

File: src/main/java/fi/aalto/cs/apluscourses/intellij/utils/CourseFileManager.java (hunk 6 above)
Comment (nikke234) [reply src/main/java/fi/aalto/cs/apluscourses/intellij/utils/CourseFileManager.java:74]: Yep thats a good idea :+1:

File: src/main/java/fi/aalto/cs/apluscourses/intellij/utils/CourseFileManager.java (hunk 8 above)
Comment (superseacat) [inline src/main/java/fi/aalto/cs/apluscourses/intellij/utils/CourseFileManager.java:47]: it seems that now most of the IO operations are `syncronized`, do you think there could be alternative options for this?

Comment (nikke234) [reply src/main/java/fi/aalto/cs/apluscourses/intellij/utils/CourseFileManager.java:46]: That's a good point. No alternatives come to mind at the moment, but we should keep it mind for the future :+1:  Luckily this is rarely used method.

File: src/test/java/fi/aalto/cs/apluscourses/presentation/MainViewModelUpdaterTest.java (hunk 1 above)
Comment (nikke234) [reply src/test/java/fi/aalto/cs/apluscourses/presentation/MainViewModelUpdaterTest.java]: That's convenient, thanks :smile: The course file metadata method doesn't throw anymore so I "disabled" this test

**Pull request merged**